File: crates/tdbin/src/verify.rs
//! Schema-independent structural verification for untrusted TDBIN bodies.

use crate::error::DecodeError;
use crate::layout::{self, WORD_BYTES};
use crate::pointer::{self, Pointer};
use crate::MAX_DEPTH;

/// Bits in one TDBIN word.
const WORD_BITS: usize = WORD_BYTES * 8;

/// Validated composite-list tag fields.
#[derive(Clone, Copy)]
struct CompositeTag {
    /// Number of inline elements.
    count: usize,
    /// Data words per element.
    data_words: u16,
    /// Pointer words per element.
    ptr_words: u16,
    /// Total element-body words declared by the list pointer.
    body_words: usize,
}

/// Verify every reachable pointer and enforce traversal limits.
pub(crate) fn message(bytes: &[u8]) -> Result<(), DecodeError> {
    let words = bytes.len() / WORD_BYTES;
    let mut budget = words.checked_sub(1).ok_or(DecodeError::BadLength)?;
    verify_pointer_word(bytes, 0, MAX_DEPTH.saturating_add(1), &mut budget)
}

/// Decode and verify one pointer word.
fn verify_pointer_word(
    bytes: &[u8],
    at: usize,
    depth: u32,
    budget: &mut usize,
) -> Result<(), DecodeError> {
    let word = layout::read_word(bytes, at)?;
    verify_pointer(bytes, at, pointer::decode(word)?, depth, budget)
}

/// Verify one already-decoded pointer.
fn verify_pointer(
    bytes: &[u8],
    at: usize,
    pointer: Pointer,
    depth: u32,
    budget: &mut usize,
) -> Result<(), DecodeError> {
    match pointer {
        Pointer::Null => Ok(()),
        Pointer::Struct {
            offset,
            data_words,
            ptr_words,
        } => verify_struct(bytes, at, offset, data_words, ptr_words, depth, budget),
        Pointer::List {
            offset,
            elem,
            count,
        } => verify_list(bytes, at, offset, elem, count, depth, budget),
    }
}

/// Verify a struct target and every pointer slot it declares.
fn verify_struct(
    bytes: &[u8],
    at: usize,
    offset: i64,
    data_words: u16,
    ptr_words: u16,
    depth: u32,
    budget: &mut usize,
) -> Result<(), DecodeError> {
    let next_depth = descend(depth)?;
    let target = target(at, offset)?;
    let words = section_words(data_words, ptr_words)?;
    require_words(bytes, target, words)?;
    consume(budget, words)?;
    verify_struct_pointers(bytes, target, data_words, ptr_words, next_depth, budget)
}

/// Verify the pointer section of one in-bounds struct body.
fn verify_struct_pointers(
    bytes: &[u8],
    target: usize,
    data_words: u16,
    ptr_words: u16,
    depth: u32,
    budget: &mut usize,
) -> Result<(), DecodeError> {
    let start = target
        .checked_add(usize::from(data_words))
        .ok_or(DecodeError::LimitExceeded)?;
    for slot in 0..usize::from(ptr_words) {
        let at = start.checked_add(slot).ok_or(DecodeError::LimitExceeded)?;
        verify_pointer_word(bytes, at, depth, budget)?;
    }
    Ok(())
}

/// Verify a list target according to its element kind.
fn verify_list(
    bytes: &[u8],
    at: usize,
    offset: i64,
    elem: u8,
    count: u32,
    depth: u32,
    budget: &mut usize,
) -> Result<(), DecodeError> {
    let next_depth = descend(depth)?;
    let target = target(at, offset)?;
    match elem {
        0 => verify_flat_list(bytes, target, 0, budget),
        1 => verify_flat_list(bytes, target, words_for_bits(count, 1)?, budget),
        2 => verify_flat_list(bytes, target, words_for_bits(count, 8)?, budget),
        3 => verify_flat_list(bytes, target, words_for_bits(count, 16)?, budget),
        4 => verify_flat_list(bytes, target, words_for_bits(count, 32)?, budget),
        5 => verify_flat_list(bytes, target, count_usize(count)?, budget),
        6 => verify_pointer_list(bytes, target, count, next_depth, budget),
        7 => verify_composite_list(bytes, target, count, next_depth, budget),
        _ => Err(DecodeError::PointerKindMismatch),
    }
}

/// Verify and charge a primitive list body.
fn verify_flat_list(
    bytes: &[u8],
    target: usize,
    words: usize,
    budget: &mut usize,
) -> Result<(), DecodeError> {
    require_words(bytes, target, words)?;
    consume(budget, words)
}

/// Verify a pointer-list body and each element pointer.
fn verify_pointer_list(
    bytes: &[u8],
    target: usize,
    count: u32,
    depth: u32,
    budget: &mut usize,
) -> Result<(), DecodeError> {
    let words = count_usize(count)?;
    require_words(bytes, target, words)?;
    consume(budget, words)?;
    for slot in 0..words {
        let at = target.checked_add(slot).ok_or(DecodeError::LimitExceeded)?;
        verify_pointer_word(bytes, at, depth, budget)?;
    }
    Ok(())
}

/// Verify a composite-list range and tag.
fn verify_composite_list(
    bytes: &[u8],
    tag_at: usize,
    elem_words: u32,
    depth: u32,
    budget: &mut usize,
) -> Result<(), DecodeError> {
    let body_words = count_usize(elem_words)?;
    let total_words = body_words
        .checked_add(1)
        .ok_or(DecodeError::LimitExceeded)?;
    require_words(bytes, tag_at, total_words)?;
    consume(budget, total_words)?;
    let tag = pointer::decode(layout::read_word(bytes, tag_at)?)?;
    verify_composite_tag(bytes, tag_at, body_words, tag, depth, budget)
}

/// Decode validated composite tag fields and verify its inline elements.
fn verify_composite_tag(
    bytes: &[u8],
    tag_at: usize,
    body_words: usize,
    tag: Pointer,
    depth: u32,
    budget: &mut usize,
) -> Result<(), DecodeError> {
    let Pointer::Struct {
        offset,
        data_words,
        ptr_words,
    } = tag
    else {
        return Err(DecodeError::MalformedCompositeTag);
    };
    let count = usize::try_from(offset).map_err(|_| DecodeError::MalformedCompositeTag)?;
    let tag = CompositeTag {
        count,
        data_words,
        ptr_words,
        body_words,
    };
    verify_composite_items(bytes, tag_at, tag, depth, budget)
}

/// Verify every inline element's pointer section.
fn verify_composite_items(
    bytes: &[u8],
    tag_at: usize,
    tag: CompositeTag,
    depth: u32,
    budget: &mut usize,
) -> Result<(), DecodeError> {
    let stride = section_words(tag.data_words, tag.ptr_words)?;
    let expected = stride
        .checked_mul(tag.count)
        .ok_or(DecodeError::LimitExceeded)?;
    if expected != tag.body_words || (stride == 0 && tag.count != 0) {
        return Err(DecodeError::MalformedCompositeTag);
    }
    let first = tag_at.checked_add(1).ok_or(DecodeError::LimitExceeded)?;
    for index in 0..tag.count {
        let offset = stride
            .checked_mul(index)
            .ok_or(DecodeError::LimitExceeded)?;
        let target = first
            .checked_add(offset)
            .ok_or(DecodeError::LimitExceeded)?;
        verify_struct_pointers(bytes, target, tag.data_words, tag.ptr_words, depth, budget)?;
    }
    Ok(())
}

/// Convert an element count and bit width to a rounded-up word count.
fn words_for_bits(count: u32, bits: usize) -> Result<usize, DecodeError> {
    count_usize(count)?
        .checked_mul(bits)
        .map(|total| total.div_ceil(WORD_BITS))
        .ok_or(DecodeError::LimitExceeded)
}

/// Convert a wire count to the platform index type.
fn count_usize(count: u32) -> Result<usize, DecodeError> {
    usize::try_from(count).map_err(|_| DecodeError::LimitExceeded)
}

/// Return a struct's total section width.
fn section_words(data_words: u16, ptr_words: u16) -> Result<usize, DecodeError> {
    usize::from(data_words)
        .checked_add(usize::from(ptr_words))
        .ok_or(DecodeError::LimitExceeded)
}

/// Resolve one relative pointer target.
fn target(at: usize, offset: i64) -> Result<usize, DecodeError> {
    layout::target(at, offset).ok_or(DecodeError::PointerOutOfBounds { word_index: at })
}

/// Require a word range to fit in the message.
fn require_words(bytes: &[u8], at: usize, words: usize) -> Result<(), DecodeError> {
    let end = at.checked_add(words).ok_or(DecodeError::LimitExceeded)?;
    (end <= bytes.len() / WORD_BYTES)
        .then_some(())
        .ok_or(DecodeError::PointerOutOfBounds { word_index: at })
}

/// Charge traversed words to the shared amplification budget.
fn consume(budget: &mut usize, words: usize) -> Result<(), DecodeError> {
    *budget = budget
        .checked_sub(words)
        .ok_or(DecodeError::AmplificationExceeded)?;
    Ok(())
}

/// Decrement the pointer traversal depth.
fn descend(depth: u32) -> Result<u32, DecodeError> {
    depth.checked_sub(1).ok_or(DecodeError::DepthExceeded)
}

File: packages/typediagram/src/tdbin/verify.ts
import { ok, type Result } from "../result.js";
import { tdbinErr } from "./error.js";
import { decodePointer, targetWord } from "./pointer.js";
import type { Pointer, TdbinError } from "./types.js";
import { WORD_BITS, WORD_BYTES, readWord, requireWordRange } from "./word.js";

const MAX_DEPTH = 64;

interface Budget {
  value: number;
}

export const verifyMessage = (bytes: Uint8Array, view: DataView): Result<void, TdbinError> => {
  const budget = { value: bytes.length / WORD_BYTES - 1 };
  return verifyPointerWord(bytes, view, 0, MAX_DEPTH + 1, budget);
};

const verifyPointerWord = (
  bytes: Uint8Array,
  view: DataView,
  at: number,
  depth: number,
  budget: Budget
): Result<void, TdbinError> => {
  const word = readWord(bytes, view, at);
  const pointer = word.ok ? decodePointer(word.value) : word;
  return pointer.ok ? verifyPointer(bytes, view, at, pointer.value, depth, budget) : pointer;
};

const verifyPointer = (
  bytes: Uint8Array,
  view: DataView,
  at: number,
  pointer: Pointer,
  depth: number,
  budget: Budget
): Result<void, TdbinError> => {
  if (pointer.kind === "null") {
    return ok(undefined);
  }
  return pointer.kind === "struct"
    ? verifyStruct(bytes, view, at, pointer, depth, budget)
    : verifyList(bytes, view, at, pointer, depth, budget);
};

const verifyStruct = (
  bytes: Uint8Array,
  view: DataView,
  at: number,
  pointer: Extract<Pointer, { kind: "struct" }>,
  depth: number,
  budget: Budget
): Result<void, TdbinError> => {
  const nextDepth = descend(depth);
  const target = targetWord(at, pointer.offset);
  const words = pointer.dataWords + pointer.ptrWords;
  const bounds = target.ok ? requireWordRange(bytes, target.value, words) : target;
  const consumed = bounds.ok ? consume(budget, words) : bounds;
  return target.ok && consumed.ok && nextDepth.ok
    ? verifyStructPointers(bytes, view, target.value, pointer.dataWords, pointer.ptrWords, nextDepth.value, budget)
    : firstError(target, consumed, nextDepth);
};

const verifyStructPointers = (
  bytes: Uint8Array,
  view: DataView,
  target: number,
  dataWords: number,
  ptrWords: number,
  depth: number,
  budget: Budget
): Result<void, TdbinError> => {
  for (let slot = 0; slot < ptrWords; slot += 1) {
    const verified = verifyPointerWord(bytes, view, target + dataWords + slot, depth, budget);
    if (!verified.ok) {
      return verified;
    }
  }
  return ok(undefined);
};

const verifyList = (
  bytes: Uint8Array,
  view: DataView,
  at: number,
  pointer: Extract<Pointer, { kind: "list" }>,
  depth: number,
  budget: Budget
): Result<void, TdbinError> => {
  const nextDepth = descend(depth);
  const target = targetWord(at, pointer.offset);
  if (!target.ok || !nextDepth.ok) {
    return firstError(target, nextDepth);
  }
  return pointer.elem === 6
    ? verifyPointerList(bytes, view, target.value, pointer.count, nextDepth.value, budget)
    : pointer.elem === 7
      ? verifyCompositeList(bytes, view, target.value, pointer.count, nextDepth.value, budget)
      : verifyFlatList(bytes, target.value, flatListWords(pointer.elem, pointer.count), budget);
};

const verifyFlatList = (
  bytes: Uint8Array,
  target: number,
  words: Result<number, TdbinError>,
  budget: Budget
): Result<void, TdbinError> => {
  const bounds = words.ok ? requireWordRange(bytes, target, words.value) : words;
  return words.ok && bounds.ok ? consume(budget, words.value) : firstError(words, bounds);
};

const verifyPointerList = (
  bytes: Uint8Array,
  view: DataView,
  target: number,
  count: number,
  depth: number,
  budget: Budget
): Result<void, TdbinError> => {
  const bounds = requireWordRange(bytes, target, count);
  const consumed = bounds.ok ? consume(budget, count) : bounds;
  if (!consumed.ok) {
    return consumed;
  }
  for (let slot = 0; slot < count; slot += 1) {
    const verified = verifyPointerWord(bytes, view, target + slot, depth, budget);
    if (!verified.ok) {
      return verified;
    }
  }
  return ok(undefined);
};

const verifyCompositeList = (
  bytes: Uint8Array,
  view: DataView,
  tagAt: number,
  bodyWords: number,
  depth: number,
  budget: Budget
): Result<void, TdbinError> => {
  const bounds = requireWordRange(bytes, tagAt, bodyWords + 1);
  const consumed = bounds.ok ? consume(budget, bodyWords + 1) : bounds;
  const word = consumed.ok ? readWord(bytes, view, tagAt) : consumed;
  const tag = word.ok ? decodePointer(word.value) : word;
  return tag.ok && tag.value.kind === "struct"
    ? verifyCompositeTag(bytes, view, tagAt, bodyWords, tag.value, depth, budget)
    : tag.ok
      ? tdbinErr("MalformedCompositeTag")
      : tag;
};

const verifyCompositeTag = (
  bytes: Uint8Array,
  view: DataView,
  tagAt: number,
  bodyWords: number,
  tag: Extract<Pointer, { kind: "struct" }>,
  depth: number,
  budget: Budget
): Result<void, TdbinError> => {
  const stride = tag.dataWords + tag.ptrWords;
  const valid = tag.offset >= 0 && stride * tag.offset === bodyWords && (stride !== 0 || tag.offset === 0);
  if (!valid) {
    return tdbinErr("MalformedCompositeTag");
  }
  for (let index = 0; index < tag.offset; index += 1) {
    const target = tagAt + 1 + stride * index;
    const verified = verifyStructPointers(bytes, view, target, tag.dataWords, tag.ptrWords, depth, budget);
    if (!verified.ok) {
      return verified;
    }
  }
  return ok(undefined);
};

const flatListWords = (elem: number, count: number): Result<number, TdbinError> => {
  const bits = [0, 1, 8, 16, 32, 64][elem];
  const total = bits === undefined ? Number.NaN : bits * count;
  return Number.isSafeInteger(total) && total >= 0 ? ok(Math.ceil(total / WORD_BITS)) : tdbinErr("PointerKindMismatch");
};

const consume = (budget: Budget, words: number): Result<void, TdbinError> => {
  if (words > budget.value) {
    return tdbinErr("AmplificationExceeded");
  }
  budget.value -= words;
  return ok(undefined);
};

const descend = (depth: number): Result<number, TdbinError> => (depth > 0 ? ok(depth - 1) : tdbinErr("DepthExceeded"));

const firstError = (...results: readonly Result<unknown, TdbinError>[]): Result<never, TdbinError> => {
  const failure = results.find((result) => !result.ok);
  return failure ?? tdbinErr("LimitExceeded");
};
